synapse/lib/view.py

```diff
@@ -1685,36 +2141,301 @@ async def delete(self):
+
+        return tagnode
+
+    async def getNodeByBuid(self, buid, tombs=False):
+        '''
+        Retrieve a node tuple by binary id.
+
+        Args:
+            buid (bytes): The binary ID for the node.
+
+        Returns:
+            Optional[s_node.Node]: The node object or None.
+
+        '''
+        nid = self.core.getNidByBuid(buid)
+        if nid is None:
+            return None
+
+        return await self._joinStorNode(nid, tombs=tombs)
+
+    async def getNodeByNid(self, nid, tombs=False):
+        return await self._joinStorNode(nid, tombs=tombs)
+
+    async def getNodeByNdef(self, ndef):
+        '''
+        Return a single Node by (form,valu) tuple.
+
+        Args:
+            ndef ((str,obj)): A (form,valu) ndef tuple.  valu must be
+            normalized.
+
+        Returns:
+            (synapse.lib.node.Node): The Node or None.
+        '''
+        buid = s_common.buid(ndef)
+        return await self.getNodeByBuid(buid)
+
+    async def _joinStorNode(self, nid, tombs=False):
+
+        node = self.livenodes.get(nid)
+        if node is not None:
+            await asyncio.sleep(0)
+
+            if not tombs and not node.hasvalu():
+                return None
+            return node
+
+        soderefs = []
+        for layr in self.layers:
+            sref = layr.genStorNodeRef(nid)
+            if tombs is False:
+                if sref.sode.get('antivalu') is not None:
+                    return None
+                elif sref.sode.get('valu') is not None:
+                    tombs = True
 
-        async with await self.snap(user=user) as snap:
-            await snap.saveNodeEdits(edits, meta=meta)
+            soderefs.append(sref)
+
+        return await self._joinSodes(nid, soderefs)
+
+    async def _joinSodes(self, nid, soderefs):
+
+        node = self.livenodes.get(nid)
+        if node is not None:
+            await asyncio.sleep(0)
+            return node
+
+        ndef = None
+        # make sure at least one layer has the primary property
+        for envl in soderefs:
+            valt = envl.sode.get('valu')
+            if valt is not None:
+                ndef = (envl.sode.get('form'), valt[0])
+                break
+
+        if ndef is None:
+            await asyncio.sleep(0)
+            return None
+
+        node = s_node.Node(self, nid, ndef, soderefs)
+
+        self.livenodes[nid] = node
+        self.nodecache.append(node)
+
+        await asyncio.sleep(0)
+        return node
 
     async def addNodeEdits(self, edits, meta):
         '''
         A telepath compatible way to apply node edits to a view.
 
         NOTE: This does cause trigger execution.
         '''
+        await self.reqValid()
+
         user = await self.core.auth.reqUser(meta.get('user'))
-        async with await self.snap(user=user) as snap:
-            # go with the anti-pattern for now...
-            await snap.saveNodeEdits(edits, None)
+
+        # go with the anti-pattern for now...
```

Comment:
Why are we calling this an antipattern? Can this comment be elaborated on, addressed or removed?

Reply:
I'm not entirely sure what this comment was referring to, but I think we should remove some of these APIs for directly saving node edits.